### PR TITLE
Dual Orbit Simulation

### PR DESCRIPTION
**Pull request by kylekrol**
Small PR to add a simulation where we model two satellites and their orbital mechanics. No attitude dynamics or attitude related actuators are included.

_This small set of changes steps through the infrastructure steps to create a new simulation and make it available in Python so it'd be very beneficial to look through what's going on here!_

### CODE REVIEW

File: src/psim/simulations/single_orbit.cpp [hunk 1]
```diff
@@ -41,11 +41,13 @@ SingleOrbitGnc::SingleOrbitGnc(Configuration const &config) {
...
-  // Orbital dynamics and lazy transformations
+  // Orbital dynamics
   add<OrbitGncEci>(config, "truth", "leader");
   add<TransformPositionEci>(config, "truth", "truth.leader.orbit.r");
   add<TransformVelocityEci>(config, "truth", "leader", "truth.leader.orbit.v");
-  // Environmental variables with body frame independent lazy transformations
+  // Environmental models
   add<EnvironmentGnc>(config, "truth", "leader");
+  add<TransformPositionEcef>(config, "truth", "truth.leader.environment.b");
+  add<TransformPositionEci>(config, "truth", "truth.leader.environment.s");
```
Comment (kylekrol) [inline src/psim/simulations/single_orbit.cpp]: These two lines here make the following transformations:

 * `truth.leader.environment.b` into `truth.eader.environment.b.eci` and `truth.eader.environment.b.ecef`
 * `truth.eader.environment.s` into `truth.eader.environment.s.eci` and `truth.eader.environment.s.ecef`

File: python/psim/sims.py [hunk 2]
```diff
@@ -2,5 +2,6 @@
 """
 
 from _psim import (
+    DualOrbitGnc,
```
Comment (kylekrol) [inline python/psim/sims.py:5]: Here is where we make the new simulation type available for use in Python.

File: python/psim/_psim/simulations.cpp [hunk 3]
```diff
@@ -101,4 +102,5 @@ static void py_assign(psim::StateFieldWritableBase &field, T const &value) {
 
 void py_simulation(py::module &m) {
   PY_SIMULATION(SingleOrbitGnc);
+  PY_SIMULATION(DualOrbitGnc);
```
Comment (kylekrol) [inline python/psim/_psim/simulations.cpp:116]: Wrapping the new model as a simulation available in Python.

File: src/psim/simulations/dual_orbit.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+#include <psim/truth/orbit.hpp>
+#include <psim/truth/time.hpp>
+#include <psim/truth/transform_position.hpp>
+#include <psim/truth/transform_velocity.hpp>
+
+#include <string>
+
+namespace psim {
+
+DualOrbitGnc::DualOrbitGnc(Configuration const &config) {
```
Comment (tanishqaggarwal) [inline src/psim/simulations/dual_orbit.cpp:39]: Seems like it should be possible to make this code reusable for the SingleOrbit case. As sensor models, etc. develop I'm worried about a lot of code copying occurring across the single and dual-orbit GNC.

**Review by tanishqaggarwal — CHANGES_REQUESTED** [review]

File: src/psim/simulations/dual_orbit.cpp (hunk 4 above)
Comment (kylekrol) [reply src/psim/simulations/dual_orbit.cpp:39]: This is a good point. I'll break it out into a `SatelliteOrbitGnc` model that's shared across both cases.

File: python/psim/_psim/simulations.cpp [hunk 5]
```diff
@@ -72,18 +73,29 @@ static void py_assign(psim::StateFieldWritableBase &field, T const &value) {
...
+template <class C>
+class PySimulation : public psim::Simulation {
+ public:
+  PySimulation() = delete;
+  virtual ~PySimulation() = default;
+
+  template <typename... Ts>
+  PySimulation(psim::Configuration const &config, Ts &&... ts)
+    : Simulation(std::make_unique<C>(config, std::forward<Ts>(ts)...)) { }
+};
```
Comment (kylekrol) [inline python/psim/_psim/simulations.cpp:85]: This allows different simulations to be exposed as different objects in Python.

**Review by tanishqaggarwal — APPROVED** [review]
Looks like a solid restructure.

**Pull request merged**